mmhuman3d/core/visualization/renderer/torch3d_renderer/uv_renderer.py

```diff
@@ -0,0 +1,537 @@
+        displacement: torch.Tensor = None,
+        displacement_map: torch.Tensor = None,
+    ) -> Meshes:
+        """Offset a vertex displacement or displacement_map to the input
+        meshes.
+
+        Args:
+            meshes (Meshes): the input meshes.
+            displacement (torch.Tensor, optional): vertex displacement.
+                shape should be (N, V, 3).
+                Defaults to None.
+            displacement_map (torch.Tensor, optional): displacement_map,
+                shape should be (N, H, W, 3).
+                Defaults to None.
+
+        Returns:
+            Meshes: returned meshes.
+        """
+        if displacement_map is not None and displacement is None:
+            displacement = self.vertex_resample(displacement_map)
+        elif displacement_map is not None and displacement is not None:
+            displacement_map = None
+        elif displacement_map is None and displacement is None:
+            warnings.warn('Redundant input, will only take displacement.')
+        batch_size = len(meshes)
+        if displacement.ndim == 2:
+            displacement = displacement[None]
+        assert displacement.shape[1] == self.NUM_VERTS
+        assert displacement.shape[0] in [batch_size, 1]
+
+        if displacement.shape[0] == 1:
+            displacement = displacement.repeat(batch_size, 1, 1)
+        C = displacement.shape[-1]
+        if C == 1:
+            displacement = meshes.verts_normals_padded() * displacement
+
+        displacement = padded_to_packed(displacement)
+
+        meshes = meshes.to(self.device)
+        meshes = meshes.offset_verts(displacement)
+        return meshes
+
+    def wrap_texture(self,
+                     meshes: Meshes,
+                     texture_map: torch.Tensor,
+                     resolution: Optional[Iterable[int]] = None,
+                     mode: Optional[str] = 'bicubic',
+                     is_bgr: bool = True) -> Meshes:
+        """Wrap a texture map to the input meshes.
+
+        Args:
+            meshes (Meshes): the input meshes.
+            texture_map (torch.Tensor): the texture map to be wrapped.
+                Shape should be (N, H, W, 3)
+            resolution (Optional[Iterable[int]], optional): resolution to
+                override self.resolution. If None, will use self.resolution.
+                Defaults to None.
+            mode (Optional[str], optional): interpolate mode.
+                Should be in ['nearest', 'bilinear', 'trilinear', 'bicubic',
+                'area'].
+                Defaults to 'bicubic'.
+            is_bgr (bool, optional): Whether the color channel is BGR.
+                Defaults to True.
+
+        Returns:
+            Meshes: returned meshes.
+        """
+
+        batch_size = len(meshes)
+        if texture_map.ndim == 3:
+            texture_map_padded = texture_map[None]
+        else:
+            texture_map_padded = texture_map
+        assert texture_map.shape[-1] == 3
```

Comment:
What about assert before padding. It needs only the original input.

Reply:
> What about assert before padding. It needs only the original input.

ok